src/index.ts

```diff
@@ -236,82 +232,121 @@ async function catchupAndWatchChain(
     });
 
     chainLogger.info("catching up with blockchain events");
-    const catchupSentinel = new AsyncSentinel();
 
     const indexerLogger = chainLogger.child({ subsystem: "DataUpdater" });
 
     // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-call
     const throttledLogProgress = throttle(
       5000,
       (currentBlock: number, lastBlock: number, pendingEventsCount: number) => {
+        const progressPercentage = ((currentBlock / lastBlock) * 100).toFixed(
+          1
+        );
+
         indexerLogger.info(
-          `pending events: ${pendingEventsCount} (indexing blocks ${currentBlock}-${lastBlock})`
+          `${currentBlock}/${lastBlock} indexed (${progressPercentage}%) (pending events: ${pendingEventsCount})`
         );
       }
     );
 
-    const indexer = await createIndexer(
-      rpcProvider,
-      storage,
-      async (indexer: Indexer<JsonStorage>, event: ChainsauceEvent) => {
-        try {
-          return await handleEvent(event, {
-            chainId: config.chain.id,
-            db: storage,
-            subscribe: (...args) => indexer.subscribe(...args),
-            ipfsGet: cachedIpfsGet,
-            priceProvider,
-            logger: indexerLogger,
-          });
-        } catch (err) {
-          indexerLogger.warn({
-            msg: "skipping event due to error while processing",
-            err,
-            event,
-          });
+    const eventHandlerContext: EventHandlerContext = {
+      chainId: config.chain.id,
+      db: storage,
+      ipfsGet: cachedIpfsGet,
+      priceProvider,
+      logger: indexerLogger,
+    };
+
+    // the chainsauce cache is used to cache events and contract reads
+    const chainsauceCache = config.cacheDir
+      ? createSqliteCache(path.join(config.cacheDir, "chainsauceCache.db"))
+      : null;
+
+    // the subscription store is used to keep track of indexed events for resuming
+    // indexing after a restart, it goes with the chain data
+    const subscriptionStore = createSqliteSubscriptionStore(
+      path.join(CHAIN_DIR_PATH, "subscriptions.db")
+    );
+
+    const indexer = createIndexer({
+      contracts: abis,
+      chain: {
+        id: config.chain.id,
+        pollingIntervalMs: 20 * 1000,
+        rpcClient: createHttpRpcClient({
+          retryDelayMs: 1000,
+          maxConcurrentRequests: 10,
+          maxRetries: 3,
+          url: config.chain.rpc,
+        }),
+      },
+      context: eventHandlerContext,
+      subscriptionStore,
+      cache: chainsauceCache,
+      logLevel: "trace",
+      logger: (level, data: unknown, message?: string) => {
+        if (level === "error") {
+          indexerLogger.error(data, message);
```

Comment:
I'm not sure whether pino methods accepts multiple arguments. If they don't, then this should work:

```ts
indexerLogger.error({ msg: message, data })
```

Reply:
this seems is working as intended at the moment, but I'll double check the arguments passed

Reply:
changed to use the object interface, it's clearer as well